### PR TITLE
Feature/acount service permissions

### PR DESCRIPTION
**Pull request by ja573**
Restructure authentication completely in the APP. Key changes include:
- [APP] move all authentication related api actions to `AccountService` (remove the `fetch!` macro from model).
- [API] login requests return `AccountDetails` instead of just the token (`LoginSession` has been removed)
- [API] token is now included in `AccountDetails`
- [APP] `RootComponent` now contains `current_user` information and provides callbacks to its child components to change the state (e.g. `Logout` callback provided to `NavbarComponent`), as well as `current_user` is now passed as properties to `AdminComponent` and `NavbarComponent` for authentication checks
- [APP] `SessionTimer` no longer contains logic to update the token, it expects a callback as its start input instead

### CODE REVIEW

File: thoth-app/src/component/admin.rs [hunk 1]
```diff
@@ -1,10 +1,11 @@
+use thoth_api::account::model::AccountDetails;
```
Comment (rhigman) [inline thoth-app/src/component/admin.rs:1]: Is it OK to borrow the API's data structure like this, or would it be better to create an app-specific version (as we have done in other similar cases)?

File: thoth-app/src/component/root.rs [hunk 2]
```diff
@@ -1,77 +1,173 @@
...
+            }
+            Msg::RenewToken => {
+                let task = self
+                    .account_service
+                    .renew_token(self.renew_token_response.clone());
+                self.renew_token_task = Some(task);
+            }
+            Msg::CurrentUserResponse(Ok(account_details)) => {
+                self.link.send_message(Msg::Login(account_details));
+                self.current_user_task = None;
```
Comment (rhigman) [inline thoth-app/src/component/root.rs:93]: Could this (and the one in the `Msg::RenewTokenResponse` case) be moved into the `Msg::Logout` case?

File: src/server/api.rs [hunk 3]
```diff
@@ -90,32 +87,43 @@ async fn login_credentials(
...
 #[post("/account/token/renew")]
 async fn login_session(
-    payload: web::Json<LoginSession>,
     token: DecodedToken,
     id: Identity,
     pool: web::Data<PgPool>,
 ) -> Result<HttpResponse, Error> {
-    let r = payload.into_inner();
-    token.jwt.as_ref().ok_or(ThothError::Unauthorised)?;
+    let email = match id.identity() {
```
Comment (rhigman) [inline src/server/api.rs:106]: Identical check performed in `account_details` below - could this be commonised?

File: thoth-app/src/service/account.rs [hunk 4]
```diff
@@ -47,8 +62,110 @@ impl AccountService {
...
+                if meta.status.is_success() {
+                    let data: Result<T, _> = serde_json::from_str(&data);
+                    if let Ok(data) = data {
+                        callback.emit(Ok(data))
+                    } else {
+                        callback.emit(Err(AccountError::ResponseError))
+                    }
+                } else {
+                    match meta.status.as_u16() {
+                        401 => callback.emit(Err(AccountError::AuthenticationError)),
```
Comment (rhigman) [inline thoth-app/src/service/account.rs:113]: Consider making constants for these HTTP status codes

**Review by rhigman — CHANGES_REQUESTED** [review]
This looks good and is working well, with the issues noted under #180 no longer present. There was some slightly odd behaviour which I describe below, although perhaps not worth tackling under this PR if there's no obvious cause/fix:
As an edge case test, I logged in, manually deleted `thoth.token` from storage, and then attempted to continue using the app. I was still able to browse data (although not edit it, as this was prevented by checks in the API) until the next scheduled `renew` request, at which point I was sent back to the login screen. That is to say, the API was sending full responses even though the POST requests had no `Authorization` header.
Some of the times I did this, the app became extremely slow and unresponsive (and my CPU went into overdrive!), even after logging back in properly. Unfortunately I couldn't work out why this only happened some times and not others, although it might have been related to still having a `sessionToken` cookie stored.

File: thoth-app/src/component/admin.rs (hunk 1 above)
Comment (ja573) [reply thoth-app/src/component/admin.rs:1]: We're currently doing both. I think in the long run we want to reuse as many structs from the API as possible, though there are many examples in which that is not possible

File: src/server/api.rs (hunk 3 above)
Comment (ja573) [reply src/server/api.rs:106]: Yes, definitely. I want to have the checked moved to an implementation of the `actix_web::FromRequest` trait down the line

File: thoth-app/src/component/root.rs (hunk 2 above)
Comment (ja573) [reply thoth-app/src/component/root.rs:93]: Not quite, otherwise we'd logout the user just after we fetch their details

File: thoth-app/src/service/account.rs (hunk 4 above)
Comment (ja573) [reply thoth-app/src/service/account.rs:113]: Good point - I'll update these

**Comment by ja573** [comment]
> This looks good and is working well, with the issues noted under #180 no longer present. There was some slightly odd behaviour which I describe below, although perhaps not worth tackling under this PR if there's no obvious cause/fix:
> As an edge case test, I logged in, manually deleted `thoth.token` from storage, and then attempted to continue using the app. I was still able to browse data (although not edit it, as this was prevented by checks in the API) until the next scheduled `renew` request, at which point I was sent back to the login screen. That is to say, the API was sending full responses even though the POST requests had no `Authorization` header.
> Some of the times I did this, the app became extremely slow and unresponsive (and my CPU went into overdrive!), even after logging back in properly. Unfortunately I couldn't work out why this only happened some times and not others, although it might have been related to still having a `sessionToken` cookie stored.

To solve this we might need to rethink the whole session management aspect of the APP. It's not clear to me how this could be solved without either keep enquiring if the token is there or requiring the token for every request to the API (even the public ones), but neither seem to be a great approach...

File: src/server/api.rs (hunk 3 above)
Comment (ja573) [reply src/server/api.rs:106]: I'm happy to tackle this now, though I'm not sure how that'll affect your changes to `Context`

File: thoth-app/src/component/root.rs (hunk 2 above)
Comment (rhigman) [reply thoth-app/src/component/root.rs:93]: Oops, meant to put this comment on the `CurrentUserResponse(Err)` case below, not this one...

Comment (rhigman) [reply thoth-app/src/component/root.rs:93]: To be clear, I meant the `self.current_user_task = None;` line only - it looked more obvious when the review was in draft!

File: src/server/api.rs (hunk 3 above)
Comment (rhigman) [reply src/server/api.rs:106]: I'm not sure either, although they're now published under #188 if that makes it clearer :) Fine to leave it if that's easiest, since changing it's already in our roadmap.

File: thoth-app/src/component/root.rs (hunk 2 above)
Comment (ja573) [reply thoth-app/src/component/root.rs:93]: Oh I see. In any case I want to separate the actions, unsetting `current_user_task` should be done when dealing with the response, rather than at logout

File: src/server/api.rs (hunk 3 above)
Comment (ja573) [reply src/server/api.rs:106]: Ok, let's try to merge the two PRs and then we'll have a separate one where we tidy up the identity checks

**Comment by rhigman** [comment]
> > This looks good and is working well, with the issues noted under #180 no longer present. There was some slightly odd behaviour which I describe below, although perhaps not worth tackling under this PR if there's no obvious cause/fix:
> > As an edge case test, I logged in, manually deleted `thoth.token` from storage, and then attempted to continue using the app. I was still able to browse data (although not edit it, as this was prevented by checks in the API) until the next scheduled `renew` request, at which point I was sent back to the login screen. That is to say, the API was sending full responses even though the POST requests had no `Authorization` header.
> > Some of the times I did this, the app became extremely slow and unresponsive (and my CPU went into overdrive!), even after logging back in properly. Unfortunately I couldn't work out why this only happened some times and not others, although it might have been related to still having a `sessionToken` cookie stored.
> 
> To solve this we might need to rethink the whole session management aspect of the APP. It's not clear to me how this could be solved without either keep enquiring if the token is there or requiring the token for every request to the API (even the public ones), but neither seem to be a great approach...

I can't see a good solution either. At least unauthorised edits are prevented, so this isn't a major problem, although the browser slowdown was somewhat bizarre - something to keep an eye on in future. Everything else is resolved now so I think this PR is fine to merge (once the branch is updated from `develop`).

**Review by rhigman — APPROVED** [review]

**Comment by ja573** [comment]
> > > This looks good and is working well, with the issues noted under #180 no longer present. There was some slightly odd behaviour which I describe below, although perhaps not worth tackling under this PR if there's no obvious cause/fix:
> > > As an edge case test, I logged in, manually deleted `thoth.token` from storage, and then attempted to continue using the app. I was still able to browse data (although not edit it, as this was prevented by checks in the API) until the next scheduled `renew` request, at which point I was sent back to the login screen. That is to say, the API was sending full responses even though the POST requests had no `Authorization` header.
> > > Some of the times I did this, the app became extremely slow and unresponsive (and my CPU went into overdrive!), even after logging back in properly. Unfortunately I couldn't work out why this only happened some times and not others, although it might have been related to still having a `sessionToken` cookie stored.
> > 
> > 
> > To solve this we might need to rethink the whole session management aspect of the APP. It's not clear to me how this could be solved without either keep enquiring if the token is there or requiring the token for every request to the API (even the public ones), but neither seem to be a great approach...
> 
> I can't see a good solution either. At least unauthorised edits are prevented, so this isn't a major problem, although the browser slowdown was somewhat bizarre - something to keep an eye on in future. Everything else is resolved now so I think this PR is fine to merge (once the branch is updated from `develop`).

Great, thanks. Re browser slowdowns I haven't been able to recreate that problem, I simply get a complain that the requests aren't authenticated - did you get any trace in the console?

**Comment by rhigman** [comment]
> > > > This looks good and is working well, with the issues noted under #180 no longer present. There was some slightly odd behaviour which I describe below, although perhaps not worth tackling under this PR if there's no obvious cause/fix:
> > > > As an edge case test, I logged in, manually deleted `thoth.token` from storage, and then attempted to continue using the app. I was still able to browse data (although not edit it, as this was prevented by checks in the API) until the next scheduled `renew` request, at which point I was sent back to the login screen. That is to say, the API was sending full responses even though the POST requests had no `Authorization` header.
> > > > Some of the times I did this, the app became extremely slow and unresponsive (and my CPU went into overdrive!), even after logging back in properly. Unfortunately I couldn't work out why this only happened some times and not others, although it might have been related to still having a `sessionToken` cookie stored.
> > > 
> > > 
> > > To solve this we might need to rethink the whole session management aspect of the APP. It's not clear to me how this could be solved without either keep enquiring if the token is there or requiring the token for every request to the API (even the public ones), but neither seem to be a great approach...
> > 
> > 
> > I can't see a good solution either. At least unauthorised edits are prevented, so this isn't a major problem, although the browser slowdown was somewhat bizarre - something to keep an eye on in future. Everything else is resolved now so I think this PR is fine to merge (once the branch is updated from `develop`).
> 
> Great, thanks. Re browser slowdowns I haven't been able to recreate that problem, I simply get a complain that the requests aren't authenticated - did you get any trace in the console?

Nothing helpful in the console, or anywhere else, unfortunately. It took me a while to reproduce it just now, and in the end it only happened after I closed and then re-opened devtools, so maybe it's a problem with the browser (Firefox 85.0.1). Hopefully it won't come up much in normal use, at any rate.
